### PR TITLE
Limit connection resources used in mapped tasks

### LINKED ISSUE #33178: Flaky `test_xcom_map_error_fails_task` test
### Body

This flaky test appears recently in our jobs and it seems this is a real problem with our code  - after few attempts of fixing it, it still appears in our builds:

```
tests/models/test_xcom_arg_map.py:174: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
airflow/utils/session.py:74: in wrapper
    return func(*args, **kwargs)
airflow/models/taskinstance.py:1840: in run
    self._run_raw_task(
airflow/utils/session.py:74: in wrapper
    return func(*args, **kwargs)
airflow/models/taskinstance.py:1494: in _run_raw_task
    session.commit()
/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/session.py:1454: in commit
    self._transaction.commit(_to_root=self.future)
/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/session.py:832: in commit
    self._prepare_impl()
/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/session.py:811: in _prepare_impl
    self.session.flush()
/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/session.py:3449: in flush
    self._flush(objects)
/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/session.py:3589: in _flush
    transaction.rollback(_capture_exception=True)
/usr/local/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py:70: in __exit__
    compat.raise_(
/usr/local/lib/python3.8/site-packages/sqlalchemy/util/compat.py:211: in raise_
    raise exception
/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/session.py:3549: in _flush
    flush_context.execute()
/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/unitofwork.py:456: in execute
    rec.execute(self)
/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/unitofwork.py:630: in execute
    util.preloaded.orm_persistence.save_obj(
/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/persistence.py:237: in save_obj
    _emit_update_statements(
/usr/local/lib/python3.8/site-packages/sqlalchemy/orm/persistence.py:1001: in _emit_update_statements
    c = connection._execute_20(
/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py:1710: in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
/usr/local/lib/python3.8/site-packages/sqlalchemy/sql/elements.py:334: in _execute_on_connection
    return connection._execute_clauseelement(
/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py:1577: in _execute_clauseelement
    ret = self._execute_context(
/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py:1953: in _execute_context
    self._handle_dbapi_exception(
/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py:2134: in _handle_dbapi_exception
    util.raise_(
/usr/local/lib/python3.8/site-packages/sqlalchemy/util/compat.py:211: in raise_
    raise exception
/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/base.py:1910: in _execute_context
    self.dialect.do_execute(
/usr/local/lib/python3.8/site-packages/sqlalchemy/engine/default.py:736: in do_execute
    cursor.execute(statement, parameters)
/usr/local/lib/python3.8/site-packages/MySQLdb/cursors.py:174: in execute
    self._discard()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <MySQLdb.cursors.Cursor object at 0x7f52bc978a60>

    def _discard(self):
        self.description = None
        self.description_flags = None
        # Django uses some member after __exit__.
        # So we keep rowcount and lastrowid here. They are cleared in Cursor._query().
        # self.rowcount = 0
        # self.lastrowid = None
        self._rows = None
        self.rownumber = None
    
        if self._result:
            self._result.discard()
            self._result = None
    
        con = self.connection
        if con is None:
            return
>       while con.next_result() == 0:  # -1 means no more data.
E       sqlalchemy.exc.ProgrammingError: (MySQLdb.ProgrammingError) (2014, "Commands out of sync; you can't run this command now")
E       [SQL: UPDATE task_instance SET pid=%s, updated_at=%s WHERE task_instance.dag_id = %s AND task_instance.task_id = %s AND task_instance.run_id = %s AND task_instance.map_index = %s]
E       [parameters: (90, datetime.datetime(2023, 8, 7, 14, 44, 7, 580365), 'test_dag', 'pull', 'test', 0)]
E       (Background on this error at: https://sqlalche.me/e/14/f405)
```

```
E       sqlalchemy.exc.ProgrammingError: (MySQLdb.ProgrammingError) (2014, "Commands out of sync; you can't run this command now")
E       [SQL: UPDATE task_instance SET pid=%s, updated_at=%s WHERE task_instance.dag_id = %s AND task_instance.task_id = %s AND task_instance.run_id = %s AND task_instance.map_index = %s]
E       [parameters: (90, datetime.datetime(2023, 8, 7, 14, 44, 7, 580365), 'test_dag', 'pull', 'test', 0)]
E       (Background on this error at: https://sqlalche.me/e/14/f405)
```

Example failures:

* https://github.com/apache/airflow/actions/runs/5786336184/job/15681127372?pr=33144

### Committer

- [X] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

### PR DESCRIPTION
**Pull request by potiuk**
The `render_template_fields` method of mapped operator needs to use database session object to render mapped fields, but it cannot get the session passed by @provide_session decorator, because it is used in derived classes and we cannot change the signature without impacting those classes.

So far it was done by creating new session in mapped_operator, but it has the drawback of creating an extra session while one is already created (remnder_template_fields is always run in the context of task run and it always has a session created already in _run_raw_task). It also causes problems in our tests where two opened database session accessed database at the same time and it cases sqlite exception on concurrent access and mysql error on running operations out of sync - likely when the same object was modified in both sessions.

This PR changes the approach - rather than creating a new session in the mapped_operator, we are retrieving the session from one stored by the _run_raw_task. It is done by context manager and adequate protection has been added to make sure that:

a) the call is made within the context manager
b) context manageer is never initialized twice in the same
   call stack

After this change, resources used by running task will be smaller, and mapped tasks will not always open 2 DB sesions.

Fixes: #33178

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by potiuk** [comment]
Hey Herel. While doing my "flaky test squashing" quest, and investigating the flaky "session" test #33178 I deduced that there must be something deep in the mapped operator implementation that causes sessions leak during the tests, that might actually be also causing similar problems in production.

It turned out that the problems are (almost for sure) caused by the way we retrieved (and not closed/discarded) DB session while rendering templates in mapped tasks. I discussed it with @uranusjr and while doing some experiments (mainly in "Chasing leaking sessions" PR https://github.com/apache/airflow/pull/33259 I realized that it has an unintended side effect that we are opening more ssessions that necessary when runnning mapped task - basically we always open one session for the "_run_raw_task" and then another session just to "render templates". 

There is a good reason for doing it (we cannot easily pass sesssion through usual @provide_session mechanism because "render_template" method's signature need to be compatible with non-mapped object, but the way we've done that was pretty reasource-heavy and brittle. It opened a new session and basically it had not closed the session, because of recursive behaviour it was quite possible the session could be close too early. 

In the discussion with @uranusjr https://github.com/apache/airflow/issues/33178#issuecomment-1674206630  it turned out that this method is only actually called in the context of `_run_raw_task` - so I figured that we can do it much more efficiently and without the "brittleness" - i.e. we could have a context manager that will store the "task" session in the global data and then render_task_template of mapped operator could use that globally stored session rather than create a new one.

I think it's much safer/stable and less resource hungry (just single session needed for the mapped task, not two) - but also far more robust (the context manager will close the session when it leaves the _run_raw_task. It's also is going to solve the test flakiness I think (i will close /reopen this PR few times to see if the problem is gone).

I cannot see any bad side-effects of the change - I already fixed all the tests that need the context manager and added the context manager to the CLI method that also uses render_templates (the only place besides "_run_raw_task" that apparently uses rendering). 

I'd love those who have a bit more experience in those parts (@ashb @uranusjr) to take a look and see if I have not missed anything - this one might be a good idea for 2.7.1 (or maybe even we could add it to rc2 of 2.7.0 if we have one).

Looking forward to comments.

**Comment by potiuk** [comment]
So far so good. No trace of session problems

**Comment by potiuk** [comment]
All passed this time. Trying again.

**Comment by potiuk** [comment]
Ok. We have one here, but it's unrelated to mapped operator (that's the first time I see it) https://github.com/apache/airflow/issues/33323  - another one to investigate.

This is the third complete test without re-occurence of the "mapped" errors - usually we would have 3-4 of those already, I consider the "mapped" as fixed but I will try few more times.

**Comment by potiuk** [comment]
All green this time. I will keep on re-running but seems that the basic problem is solved. @uranusjr - if you agree, we seem to have an RC2 for 2.7.0 is coming, do you think it's good to get it for RC2?

**Comment by potiuk** [comment]
Reopened to rebuild

**Comment by potiuk** [comment]
Ok. We had one more failure with sqlite last time, but it was in teardown, so a bit different, and I **think** it was caused by lack of passing session to create_dagrun. I added session to it, and will continue rebasing it. 

One of the followups of this one might be a global attempt to make sure in all our tests whenever ti.run() or dag_maker.create_dagrun() are called we might want to  add some pre-commit/ast to force it's usage

**Comment by potiuk** [comment]
Seems that there is yet another reason for the sessions. The quest is not over yet.

**Comment by potiuk** [comment]
cc: @uranusjr @ashb - would really love to get it in - this one seems to stabilize a lot, still there are some cases where tests are flaky but they are few and far between comparing to recently observed flakiness (+ it also saves resources for mapped tasks).

File: airflow/cli/commands/task_command.py [hunk 1]
```diff
@@ -634,6 +635,12 @@ def task_test(args, dag: DAG | None = None) -> None:
                 session.delete(ti.dag_run)
 
 
+@provide_session
+def render_template_with_session(ti: TaskInstance, session=NEW_SESSION):
+    with set_current_task_instance_session(session=session):
+        ti.render_templates()
```
Comment (uranusjr) [inline airflow/cli/commands/task_command.py]: Why does this need a separate function?

Comment (potiuk) [reply airflow/cli/commands/task_command.py]: To use decorator applired to a minimum scope it is needed. We do not need to keep session open after we already rendered the templates - i.e when we print the rendeder template values.

Comment (potiuk) [reply airflow/cli/commands/task_command.py]: I can also make it a nested function inside "task_render", but I am rather sceptical about adding functions inside functions. We could also potentially turn "provide_session" into a generic context manager but that would likely need some modifications to it and it would be introducing a new way of managing and allocating sessions - I would rather not introduce a new way of managing sessions but I would rely on how we do it elsewhere. 

I considered all those options and extracting the method the way I did, sounded like the most "straightforward" solution.

Comment (uranusjr) [reply airflow/cli/commands/task_command.py]: There’s `create_session`, a decorator is not needed.

```python
with create_session() as session, set_current_task_instance_session(session=session):
    ti.render_templates()
```

Comment (potiuk) [reply airflow/cli/commands/task_command.py]: Right.

Comment (potiuk) [reply airflow/cli/commands/task_command.py]: Fixed.

**Comment by uranusjr** [comment]
I’m having second thoughts. If this works, why can’t we just pass the `session` all the way from `_run_raw_task` into `render_template_fields` instead? The fact we could not do this previously seems to hint that using a context manager to clean this up will not work.

**Comment by potiuk** [comment]
The main reason was the signature of the render_templelate method that could not be changed for the regular non-mapped operator. 

At least that"s what the comment says in the original comment.

The regular operators do not need the session at all. And we are extending the regular operator methods. So this is the main reason why we cannot pass the session directly. And the only one.

The way it was implemented created (and not closed) a new session. It did not close it because it was  hard with the recursive patterns of template rendering.

The way I implemented it now - we so not create new session. But use the one that comes with the task - and close it when run method completes. This Is the same as creating session but without the overhead of creating a new one - just reusing one that we already have 

So basically we are doing what we've been doing already but with one session rather than two - and we are closing the session when. It is not needed. This is the saving we have.

Do you see any flaw with this reasoning? Do you think it's not worthy a saving of extra session and db connection created always when mapped operator is running? I think it is quite significant improvement TBH, without -seemingly - any drawback.

**Comment by potiuk** [comment]
The main reason was the signature of the render_templelate method that could not be changed for the regular non-mapped operator. 

At least that"s what the comment says in the original comment.

The regular operators do not need the session at all. And we are extending the regular operator methods. So this is the main reason why we cannot pass the session directly. And the only one.

The way it was implemented created (and not closed) a new session. It did not close it because it was  hard with the recursive patterns of template rendering.

The way I implemented it now - we so not create new session. But use the one that comes with the task - and close it when run method completes. This Is the same as creating session but without the overhead of creating a new one - just reusing one that we already have 

So basically we are doing what we've been doing already but with one session rather than two - and we are closing the session when. It is not needed. This is the saving we have.

Do you see any flaw with this reasoning? Do you think it's not worthy a saving of extra session and db connection created always when mapped operator is running? I think it is quite significant improvement TBH, without -seemingly - any drawback.

**Comment by potiuk** [comment]
Or maybe you think we can change the signature ? I believe (on the phone now) the original comment mentioned that the signature cannot be changed because the methods that would have to be changed are already in the wild being extended by existing operators and by changing the signature and adding session would breake those operators. 

I would very much prefer it this way as well if we can do it - do you think it is possible by looking at the comment ?

**Comment by potiuk** [comment]
@uranusjr ?

**Comment by potiuk** [comment]
Hey @uranusjr -> I would ***REALLY*** love that one to get into 2.7.1 - and basically I run it so many times (and I think the proposal is pretty sound) that I have a reason to believe it does solve some of the resourcing issues. We have a number of reports indicating that Airflow became much more resource hungry in recent versions. I am not sure if that could be one of the reason  but I am quite certain it is going to decrease the amount of resources needed by the workers. 

It just basically works for your approval - maybe we can ask others to review it, but I think it's not a good idea to keep it in the limbo forever.

**Comment by uranusjr** [comment]
The only thing I am wondering is whether the getter should return a new session if one is not set previously. Currently it would raise an error in this case, which is arguably a compatibility issue; we can probably categorise this change as an enhancement and supply a significant change notice in 2.8.0, but for 2.7.x I would prefer this to not break unexpected code paths in user code.

**Comment by potiuk** [comment]
> The only thing I am wondering is whether the getter should return a new session if one is not set previously. Currently it would raise an error in this case, which is arguably a compatibility issue; we can probably categorise this change as an enhancement and supply a significant change notice in 2.8.0, but for 2.7.x I would prefer this to not break unexpected code paths in user code.

If that's the only worry @uranusjr - I fixed it. Instead of raising errror, I print a warning with detailed stack trace and asking the user to report it.

**Comment by potiuk** [comment]
@uranusjr ?

**Review by uranusjr — APPROVED** [review]
